R/weakref.R

```diff
@@ -0,0 +1,87 @@
+#' Create a weak reference
+#'
+#' @description
+#'
+#' A weak reference is a special R object which makes it possible to keep a
+#' reference to an object without preventing garbage collection of that object.
+#' It can also be used to keep data about an object without preventing GC of the
+#' object, similar JavaScript WeakMap.
+#'
+#' As long as the key is reachable, the value will not be garbage collected.
+#' This is true even if the weak reference object becomes unreachable.
+#'
+#' When the key becomes unreachable, the key and value in the weak reference
+#' object are replaced by `NULL`, and the finalizer is scheduled to execute.
+#'
+#' @param key The key for the weak reference. Must be a reference object -- that
+#'   is, an environment or external pointer.
+#' @param value The value for the weak reference. This can be `NULL`, if you
+#'   want to use the weak reference like a weak pointer.
+#' @param finalizer A function that is run after the key becomes unreachable.
+#' @param onexit Should the finalizer be run when R exits?
```

Comment:
"onexit" is evocative of `on.exit()` handlers. Maybe `ensure`? Should it be `TRUE` by default?

Reply:
I was thinking of the `onexit` from `reg.finalizer()`, but if you think that's too confusing with the `on.exit()` function, I could change it. `ensure` seems a little weird to me though.

In my experience, it's rarely necessary to run a finalizer on exit, because I've used them for managing stuff that's all in R. But I don't feel strongly about the default either way.

Reply:
How about `on_quit`? This is consistent with `quit()`. (It needs the underscore so we don't introduce a new casing style in the UI.)

Keeping the default as `FALSE` sounds good.

Reply:
Sure, `on_quit` sounds good to me.